### ISSUE TITLE
AdditionalFiles is empty in CompilationStartAnalysisContext.Options.AdditionalFiles but correct in SyntaxNodeAnalysisContext.Options.AdditionalFiles

### ISSUE THREAD

**Issue by bichuga** [dotnet/roslyn#35357]
**Version Used**: 
Microsoft CodeAnalysis 2.8
Visual Studio 2019 Preview 2

**Description**
AdditionalFiles is empty on a CompilationStartAction but is correctly populated on a SyntaxNodeAction. This prevents users from taking action which require AdditionalFiles at compilation start.

**Steps to Reproduce**:

1. Create an analyzer which registers a CompilationStartAction

        public override void Initialize(AnalysisContext context)
        {
            context.RegisterCompilationStartAction(this.AnalyzeAdditionalFiles);
        }

        private void AnalyzeAdditionalFiles(CompilationStartAnalysisContext context)
        {
            if (!context.Options.AdditionalFiles.Any()) 
                   Debugger.Launch();
        }

2. In your consuming project, add an additional file

    <ItemGroup>
      <AdditionalFiles Include="foo.log" />
    </ItemGroup>

3. Add the analyzer to your project

**Expected Behavior**:
Debugger should not launch

**Actual Behavior**:
Debugger launches

4. Change your analyzer to register a SyntaxNodeAction instead
        
        public static readonly SyntaxKind[] PossibleSyntaxKinds = { SyntaxKind.SimpleMemberAccessExpression, SyntaxKind.ObjectCreationExpression };

        public override void Initialize(AnalysisContext context)
        {
            context.RegisterSyntaxNodeAction(this.AnalyzeSyntaxNode, PossibleSyntaxKinds);
        }

        private void AnalyzeSyntaxNode(SyntaxNodeAnalysisContext context)
        {
            if (!context.Options.AdditionalFiles.Any()) 
                   Debugger.Launch();
        }

**Expected Behavior**:
Debugger does not launch

**Actual Behavior**:
Debugger does not launch. (Behavior is correct)

Note: May be related to #12840

**Comment by davkean** [comment]
Is this a SDK-based project? This might because design-time build hasn't finished which then pushes this to Roslyn.

**Comment by bichuga** [comment]
Nope, just a vanilla project.

**Comment by mavasani** [comment]
@bichuga I am unable to repro your issue. I have attached the Analyzer project, [Analyzer1.zip](https://github.com/dotnet/roslyn/files/3132894/Analyzer1.zip), that I created with your repro steps and the target analysis solution, [ClassLibrary16.zip](https://github.com/dotnet/roslyn/files/3132895/ClassLibrary16.zip), with 2 projects:
1. SDK-style netstandard C# class library
2. csproj based .NETFramework C# class library.

Both these projects have an additional file and I verified that the `CompilationStartAnalysisContext` always gets the additional file and debugger is never launched.

I verified both with `Microsoft Visual Studio Enterprise 2019 Int Preview Version 16.1.0 Preview 2.0 [28817.7.d16.1]`, which has Microsoft.CodeAnalysis version `3.100.19.21101`, and also by locally building Roslyn.sln and deploying them to the same hive as the analyzer. I tried running your scenario with both debugger attached and no debugger attached. The `Debugger.Launch` is hit if I comment out the AdditionalFiles ItemGroup, but not hit otherwise.

**Comment by bichuga** [comment]
@mavasani I discovered that this bug repros only on the first time opening a project. If you close your solution and reopen, additionalfiles is populated properly for CompilationStart. I've updated the repro steps to reflect.

**Comment by jasonmalinowski** [comment]
@bichuga Are you not seeing a later CompilationStartAction that does have it? The repro steps as written are not really testing the right thing: when a project is loading it's perfectly "normal" to see files missing while we still load project information. Because collecting everything is expensive, we do that as asynchronously as possible.

**Comment by mavasani** [comment]
I agree with @jasonmalinowski here. The expectation for analyzer execution in IDE is to see deterministic behavior after the project/solution loading etc. has come to a steady state.

**Comment by jasonmalinowski** [comment]
And there may be a bug here too as we generally shouldn't be starting IDE analyzers until after we do have the design time build because there isn't much of a point of running them. But if a design-time build entirely failed, we still wouldn't have them and would run analyzers anyways.

**Comment by davkean** [comment]
In the repro, both the analyzer and additional file are added at the same time to Roslyn in the same batch, @jasonmalinowski why wouldn't the analyzer see this?

**Comment by bichuga** [comment]
> @bichuga Are you not seeing a later CompilationStartAction that does have it? The repro steps as written are not really testing the right thing: when a project is loading it's perfectly "normal" to see files missing while we still load project information. Because collecting everything is expensive, we do that as asynchronously as possible.

All CompilationStartActions for the lifetime of VS will have empty AdditionalFiles collection until you close and reopen VS.

**Comment by jasonmalinowski** [comment]
@davkean I guess until you say that now I couldn't assume that, since that's an implementation detail you're entitled to change.

> All CompilationStartActions for the lifetime of VS will have empty AdditionalFiles collection until you close and reopen VS.

OK, _that's_ the confirmation I wanted. That's....funky.

**Comment by davkean** [comment]
I think it's a good assumption that if Analyzer and AdditionalFiles are added at exactly the same time, either via a NuGet package or via a manual edit, then we can guarantee that Roslyn sees them at the same time. It would be up to Roslyn after that to make sure it appears in Options.AdditionalFiles at the same time as the analyzer is instantiated.

If the AdditionalFiles or the Analyzer are added at different times (say one via the Add Analyzer UI and the one via the Properties -> Build Action UI) there will be no such guarantee.

**Comment by elliotmoule** [comment]
Just adding to this, a few years later. I believe this bug still exists.

I followed the instructions for using Additional Files, as part of the [Roslyn docs](https://github.com/dotnet/roslyn/blob/25623fe1edcb6c31a4d6cc5075e7e4ac18c719ff/docs/analyzers/Using%20Additional%20Files.md), and when running my code, am not getting my additional files.

I am using an SDK style project, and read above that it may not be ready by the compilation start, but the docs don't mention that as a potential issue, even though they utilise the compilation start action. Is there a known work around? Is this just a case of potentially needing to wait for the files to populate?

**Comment by LeaFrock** [comment]
Still no progress 💢 ? I'm sucked in this problem when I want to write a source generator.

**Comment by CyrusNajmabadi** [comment]
I'm  unable to repro this, including trying out the examples above.  If you're not seeing additional files, please provide us with a repro case (ideally a github link) where that happens so we can debug through that particualr case.  It's likely misauthoring.
